src/defrag.c

```diff
@@ -1162,18 +1162,20 @@ void activeDefragCycle(void) {
                 db = NULL;
                 server.active_defrag_running = 0;
 
-                /* If the defragmentation hits and defragmented bytes are both
-                 * below 1%, which indicates low defragmentation efficiency.
-                 * Reduce the defragmentation speed by 10% for the next cycle. */
                 long long last_hits = server.stat_active_defrag_hits - start_hits;
                 long long last_misses = server.stat_active_defrag_misses - start_misses;
-                if (last_hits < (last_hits + last_misses) * 0.01 &&
-                    start_frag_bytes >= frag_bytes &&
-                    (start_frag_bytes - frag_bytes) < start_frag_bytes * 0.01)
+                float last_frag_pct_change = start_frag_pct - frag_pct;
+                /* When defragmentation efficiency is low, we gradually reduce the
+                 * speed for the next cycle to avoid CPU waste. However, in the
+                 * following two cases, we keep the normal speed:
+                 * 1) If the fragmentation percentage has decreased and defragmentation hits are above 1%.
+                 * 2) If the fragmentation percentage have increased by 5%. */
+                if ((last_frag_pct_change < 0 && last_hits >= (last_hits + last_misses) * 0.01) ||
+                    (last_frag_pct_change > 0 && last_frag_pct_change > 0.05))
```

Comment:
wasn't it logically redundant? 
```suggestion
                    (last_frag_pct_change > 0.05))
```

Reply:
If the above condition is false, last_frag_pct_change can still be negative here.
Although it's the same without `>0`, it is not consistent with the fragmentation rate increase in the comment above.